Mobile 3817

dpalou opened the pull request.

NoelDeMartin commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+                },
+                error: error => reject(error),
+                complete: () => {
+                    subscription && subscription.unsubscribe();
```
Should we use the same strategy here as in `CoreSubscription.once`? Meaning, to do this asynchronously in a microtask just in case that this is executed instantly. Also, should we unsubscribe in error as well? And finally, this is yet another example of relying on the particular behaviour of WS observables which only emit two values.

dpalou replied:
Nice catch, in error I need to unsubscribe and also consider the request as completed (once an observable has errored, the complete callback isn't called). So I created a new `complete` function that is called in both cases, and I use nextTick to make sure the subscription object is defined.

NoelDeMartin commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+     *
+     * @param component Component instance.
+     */
+    startComponentLoad(component: AsyncComponent): RequestManagerLoadCycle {
```
I'm not sure about the naming for this method, given that this doesn't acutally start anything. Maybe it should be called `componentLoadStarted`? or `watchComponent`?

Same problem with `startPageLoad`.

NoelDeMartin commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+     * @param hasMeaningfulChanges Callback to check if there are meaningful changes if data was updated in background.
+     * @return First value of the observable.
+     */
+    handleRequest<T>(
```
I think "handle" is too generic of a verb to use here, specially since the documentation doesn't give any more details as to what's going on. Maybe we should call this `watchRequest` (same as `watchComponent` & `watchPage`, if you like my suggestions), and give a bit more details on the description of what this method actually does.

NoelDeMartin commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+
+    protected initialPath: string;
+    protected currentLoadCycle?: RequestManagerLoadCycle;
+    protected pendingLoadCycles = new Set<RequestManagerLoadCycle>();
```
Instead of "load cycles", if we end up renaming this class to `PageUpdatesManager`, should this be renamed to `ongoingPageUpdates`? I prefer `ongoing` to `pending` in this context, and I think the name of the thing the manager manages should be called the same way as the manager (for example, if we keep the name `RequestManager`, this should be called `ongoingRequests` but that isn't correct).

If this changes, the class should change as well. `PageUpdate`? We can discuss naming overall instead of looking at individual parts.

dpalou replied:
I changed "pending" to "ongoing", but I kept the other names until we discuss the overall naming.

NoelDeMartin commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+ */
+export class RequestManager {
+
+    onUpdateData = new Subject<void>();
```
Given that this is triggered from pressing the `refresh-button-modal`, and this manages a page, should this be called `onRefreshPage` instead?

NoelDeMartin reviewed (changes requested):
I've found some things to improve, but I think the overall approach looks good. Also, there seems to be some conflicts with integration, so please rebase your branch.

dpalou commented on src/types/config.d.ts:
```diff
@@ -71,4 +71,6 @@ export interface EnvironmentConfig {
     uselegacycompletion?: boolean; // Whether to use legacy completion by default in all course formats.
     toastDurations: Record<ToastDuration, number>;
+    callWSInBackground?: boolean; // Whether to allow calling WS in background, displaying cached data immediately.
+    callWSInBackgroundExpirationTime?: number; // Number of ms to consider an entry expired when calling WS in background.
```
Actually, what is misleading is the name "_expirationTime_" stored in the cache. At the beginning, this value did contain the expiration time of the entry. However, this changed when we decided to have dynamic expiration times (e.g. based on if the user uses WiFi or not), and we decided not to modify the DB table because we didn't want to risk losing data if something went wrong. So now the column "_expirationTime_" actually contains the time when the cache entry was saved (is set to Date.now()), it's not the expiration time.

dpalou replied:
It shouldn't be hard in terms of code, but if something goes wrong the user would lose all cached data (and offline access). As you know, Juan always wants to be really careful with this type of things. We can discuss it internally if you want, but I'd do it in a different issue.

dpalou commented on src/core/classes/site.ts:
```diff
@@ -1466,54 +1621,64 @@ export class CoreSite {
+            }),
+        );
+
+        this.ongoingRequests[cacheId] = observable;
```
No, because this request uses the AJAX endpoint, it doesn't use the normal WS endpoint (and so it doesn't need to send the token, the response format is different, etc.). We never cache responses of AJAX calls, but this was a special case introduced in 4.0 to improve the performance of the user menu. Since this was a special case I decided to just put the logic I needed here.

dpalou commented on moodle.config.json:
```diff
@@ -109,5 +109,6 @@
         "sticky": 0
-    }
+    },
+    "callWSInBackground": true
```
As we discussed in chat, I finally changed the name to "disable" and I removed it from here.

dpalou commented on src/core/features/courses/services/courses.ts:
```diff
@@ -819,89 +931,110 @@ export class CoreCoursesProvider {
-                // Move category field to categoryid on a course.
-                course.categoryid = course.category;
-                delete course.category;
+        return this.getUserCoursesObservable({
+            readingStrategy: strategy,
+            siteId,
+        }).toPromise();
```
I saw it too while fixing another thing xD It should be _firstValueFrom_, yes.

dpalou commented on src/core/classes/site.ts:
```diff
@@ -610,172 +638,285 @@ export class CoreSite {
+     * @param wsPreSets Extra options related to the WS call.
+     * @return Observable.
+     */
+    protected performRequest<T = unknown>(
```
I created a new type called WSObservable. It'll be in a new commit because the change had conflicts with several commits, it wasn't easy to squash.

dpalou commented on src/core/utils/rxjs.ts:
```diff
@@ -60,3 +62,154 @@ export function startWithOnSubscribed<T>(onSubscribed: () => T): OperatorFunctio
+            }
+        };
+
+        // Before subscribing, initialize the data for all observables.
+        observables.forEach((observable, obsIndex) => {
+            observablesData[obsIndex] = {
+                values: [],
+                hasValueForIndex: [],
+                completed: false,
+            };
+        });
+
+        observables.forEach((observable, obsIndex) => {
+            const observableData = observablesData[obsIndex];
+
+            observableData.subscription = observable.subscribe({
+                next: (value) => {
+                    observableData.values.push(value);
+                    observableData.hasValueForIndex.push(true);
+                    treatEmitted();
+                },
+                error: (error) => {
+                    hasErrored = true;
+                    subscriber.error(error);
+                },
+                complete: () => {
+                    observableData.completed = true;
+                    numCompleted++;
+                    treatEmitted(true);
+                },
+            });
+        });
```
As we talked in chat, I need to initialize all items before treating them. I changed the _forEach_, now I use _map_ instead.

dpalou commented on src/core/utils/rxjs.ts:
```diff
@@ -60,3 +62,154 @@ export function startWithOnSubscribed<T>(onSubscribed: () => T): OperatorFunctio
+            observableData.subscription = observable.subscribe({
+                next: (value) => {
+                    observableData.values.push(value);
+                    observableData.hasValueForIndex.push(true);
```
Storing the last value is not enough, here's an example. I have observer A that emits values AX, observer B emits values BX. If they emit (in this order) A1, A2, A3, B1, I want my operator to emit [A1, B1]. If I only store the last value I'd only have A3, A1 would have been lost.

The only optimization we could do is to remove the items from a certain index once all observables have emitted a value for that index. But as you say, the chances of having a memory leak are really small so I don't think it's worth it.

dpalou commented on src/core/classes/site.ts:
```diff
@@ -2193,6 +2369,68 @@ export class CoreSite {
+ * @param callback Callback called with the result of current request and the reading strategy to use in next requests.
+ * @return Operator.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function chainRequests<T, O extends ObservableInput<any>>(
```
I put it in here because this method relies a lot on CoreSite functionality, and we don't have a CoreSite utils.

I changed the type to use the new WSObservable type.

dpalou commented on src/core/utils/rxjs.ts:
```diff
@@ -60,3 +62,154 @@ export function startWithOnSubscribed<T>(onSubscribed: () => T): OperatorFunctio
+                return;
+            }
+
+            if (numCompleted >= observables.length) {
+                subscriber.complete();
+
+                return;
+            }
```
I detected a bug in this part, if there were values pending to be emitted when the last observable completed they were never emitted. I added a new commit to fix this.

This function has changed a bit because I fixed several things you mentioned, I recommend you to review the whole function again once I upload the new code.

dpalou commented on src/core/features/courses/services/courses.ts:
```diff
@@ -553,38 +586,57 @@ export class CoreCoursesProvider {
+                'core_course_get_enrolled_courses_by_timeline_classification',
+                params,
+                preSets,
+            ).pipe(map(response => response.courses));
```
These checks were added when we didn't have types in WS, and AFAIK they were there "just in case". When I find a case like this I usually remove them, the return field is not optional so the app will always receive it.

dpalou commented on upgrade.txt:
```diff
@@ -6,6 +6,7 @@ information provided here is intended especially for developers.
 - Zoom levels changed from "normal / low / high" to " none / medium / high".
 - --addon-messages-* CSS3 variables have been renamed to --core-messages-*
 - The database constants in CoreSite (WS_CACHE_TABLE, CONFIG_TABLE, LAST_VIEWED_TABLE) and the DBRecord types have been moved to src/core/services/database.
+- The component <core-block> will no longer detect changes of properties inside the extraData object, it will only detect changes if the object itself changes.
```
I wanted to add `ngOnChanges` in the block component, and Angular doesn't recommend declaring both `ngOnChanges` and `ngDoCheck` ([they have a eslint rule for that](http://codelyzer.com/rules/no-conflicting-life-cycle-hooks/)). So IMO there were 3 options:

1- Remove the `ngDoCheck` (like I did).
2- Instead of adding `ngOnChanges`, put the new code in `ngDoCheck`. But that means the new checks will be executed way more often.
3- Ignore the eslint rule.

I think in the past we talked about limiting this and removing `ngDoCheck`, so I went for that approach. I wanted to discuss this in one of our dev meetings, but then annual leave came and I forgot.

NoelDeMartin replied:
Ok that makes sense, we can leave it as is then.

dpalou commented on src/core/features/block/classes/base-block-component.ts:
```diff
@@ -40,6 +42,7 @@ export abstract class CoreBlockBaseComponent implements OnInit, ICoreBlockCompon
 
     loaded = false; // If the component has been loaded.
     protected fetchContentDefaultError = ''; // Default error to show when loading contents.
+    protected onReadyPromise = new CorePromisedValue<void>();
```
As we talked in chat, there's a small difference between `loaded` and `onReadyPromise`, because `loaded` can be set to false to force displaying a loading in the UI, while the `onReadyPromise` shouldn't change. I'm leaving it as it was.

dpalou commented on src/core/classes/site.ts:
```diff
@@ -1704,16 +1869,27 @@ export class CoreSite {
+     * @param readingStrategy Reading strategy.
+     * @return Observable returning site config.
+     */
+    getConfigObservable(name?: undefined, readingStrategy?: CoreSitesReadingStrategy): Observable<CoreSiteConfig>;
```
I removed the third overload (this means I have to cast `name` to string in `getConfig` to avoid a TS error).

dpalou commented on src/core/features/course/services/course.ts:
```diff
@@ -391,19 +394,36 @@ export class CoreCourseProvider {
+                courseid: courseId,
+                returncontents: true,
+            };
+            const preSets: CoreSiteWSPreSets = {
+                cacheKey: this.getCourseBlocksCacheKey(courseId),
+                updateFrequency: CoreSite.FREQUENCY_RARELY,
+                ...CoreSites.getReadingStrategyPreSets(options.readingStrategy),
```
Just for the record, I think this comment should be in the `getSections` function, I guess it was placed here by mistake.

I applied your suggestion and I allow passing `preSets` to the observable function, so now `getReadingStrategyFromPreSets` isn't needed. I kept the `readingStrategy` option so places using the observable function can use it and it's coherent with other functions.

dpalou commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+/**
+ * Class to manage requests in a page and its components.
+ */
+export class RequestManager {
```
I named it `PageRequestsManager` for now, we can discuss this with the team. I also moved the other class to a new file.

dpalou commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+            closeOnNavigate: true,
+        });
+
+        this.onUpdateData.next();
```
IMO yes, the page should refresh if the user leaves (that way he'll see the latest data when he comes back).

I agree it looks a bit weird that `openModal` waits for the modal to be closed, but it's been like this for a long while, if we want to change this IMO we should do it in a different issue.

dpalou commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+    protected pendingLoadCycles = new Set<RequestManagerLoadCycle>();
+
+    constructor() {
+        this.initialPath = CoreNavigator.getCurrentPath();
```
Moved to `startPageLoad`.

dpalou commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+            return;
+        }
+
+        // It seems load cycle has finished. Wait to make sure no new component has been rendered and started loading.
+        clearTimeout(this.finishTimeout);
+        this.finishTimeout = window.setTimeout(() => {
+            // Loading finished.
+            this.requestManager.finishLoadCycle(this);
```
The timeout is cleared when a new component starts loading or a new request is done, so in those cases the timeout won't trigger, that's why it isn't needed to duplicate the if. Do you want me to specify it in the comment?

dpalou replied:
I finally added a comment to make it clearer.

dpalou commented on src/core/classes/request-manager.ts:
```diff
@@ -0,0 +1,269 @@
+     *
+     * @param loadCycle Load cycle that finished.
+     */
+    finishLoadCycle(loadCycle: RequestManagerLoadCycle): void {
```
I cannot make it protected because it's called from the other class.

I added an `if` to check if the request is no longer in the list (it means it finished before).

I wanted to check both readiness and requests to make it more robust, maybe some page/component loads the main data first (so it's ready) and then loads a certain data without blocking the main UI (and it doesn't use a sub-component to do that). Or maybe someone forgets to put an await somewhere. It's probably an uncommon use case, but since we're already handling the requests it was easy to keep a counter of the requests.

dpalou commented:
Thanks Noel, I fixed most of the things you said. The conversations that I resolved without any comment means I applied your suggestion. I squashed most changes into existing commits, and I added 2 new commits too.

NoelDeMartin commented on src/core/classes/page-load-watcher.ts:
```diff
@@ -0,0 +1,157 @@
+     * @return Whether this load watcher can update data in background.
+     */
+    canUpdateInBackground(): boolean {
+        return this.staleWhileRevalidate;
```
This is a bit weird, why not continue calling this variable `updateInBackground` or something similar? I think it's ok that the `getReadingStrategy` returns the `STALE_WHILE_REVALIDATE` strategy then.

If we want to leave it like this, I think it should be called `usesStaleWhileRevalidateStrategy` or something, but I guess that's too long.

NoelDeMartin commented on src/core/classes/page-load-watcher.ts:
```diff
@@ -0,0 +1,157 @@
+    protected hasChanges = false;
+    protected ongoingRequests = 0;
+    protected components = new Set<AsyncComponent>();
+    protected finishTimeout?: number;
```
Maybe this should be called `loadedTimeout`?

NoelDeMartin commented on src/core/classes/page-load-watcher.ts:
```diff
@@ -0,0 +1,157 @@
+                    }
+
+                    // Second value, it means data was updated in background. Compare data.
+                    if (hasMeaningfulChanges && hasMeaningfulChanges(firstValue, value)) {
```
```suggestion
                    if (hasMeaningfulChanges?.(firstValue, value)) {
```

I don't remember if I already suggested this, sorry if I did. But in case you didn't know, this is how you use optional chaining with method calls :). This can be also applied to other parts of this PR, and the codebase overall if we like it (I do?).

dpalou replied:
I think I've used it in some places, but IIRC some other dev in the team (I can't remember who it was, maybe Juan) said it looked a bit weird so I stopped using it. I'm ok with using it so I changed it, maybe we can talk about this in one of our dev meetings.

NoelDeMartin commented on src/core/classes/page-load-watcher.ts:
```diff
@@ -0,0 +1,157 @@
+    ): Promise<T> {
+        return new Promise((resolve, reject) => {
+            let subscription: Subscription | null = null;
+            let firstValue: T | undefined;
```
I think instead of declaring this, we can use `CorePromisedValue` instead. Furthermore, if we do that, we can rewrite this method without needing to declare the body within the `new Promise` initialization.

NoelDeMartin commented on src/core/classes/page-loads-manager.ts:
```diff
@@ -0,0 +1,146 @@
+     * @return Load watcher to use.
+     */
+    startComponentLoad(component: AsyncComponent): PageLoadWatcher {
+        if (!this.currentLoadWatcher) {
+            // A component is loading data without the page loading data, probably the component is reloading/refreshing.
+            // Use a load watcher instance but don't store it in currentLoadWatcher because it's not a page load.
+            const componentLoad = new PageLoadWatcher(this, false);
+            componentLoad.watchComponent(component);
+
+            return componentLoad;
+        }
+
+        this.currentLoadWatcher.watchComponent(component);
+
+        return this.currentLoadWatcher;
```
```suggestion
        // A component is loading data without the page loading data, probably the component is reloading/refreshing.
        // Use a load watcher instance but don't store it in currentLoadWatcher because it's not a page load.
        const loadWatcher = this.currentLoadWatcher ?? new PageLoadWatcher(this, false);

        loadWatcher.watchComponent(component);

        return loadWatcher;
```

NoelDeMartin commented on src/core/classes/page-load-watcher.ts:
```diff
@@ -0,0 +1,157 @@
+    protected finishTimeout?: number;
+
+    constructor(
+        protected requestsManager: PageLoadsManager,
```
I guess this should be called `loadsManager` instead, but looking at this I think it can potentially be improved further (although maybe it requires too much work :sweat_smile:).

Instead of passing the manager here, we could accept listeners for this class, and call the `onPageLoaded` callback from an interface. That way, this class will be decoupled from the PageLoadsManager (although in practice I don't think it'll be used anywhere else).

More generally, I see this type of situation quite often in the code base, for example the same issue happens in the PageLoadsManager itself, and we're solving it using a Subject for `onRefreshPage`. Maybe we should roll out a solution for classes allowing listeners and use the same everywhere? I guess we could rely on the Observable API, but personally I prefer a traditional Listener approach.

dpalou replied:
I changed the name to `loadsManager`, but I haven't changed the implementation because, as you said, I don't think this class will be used anywhere else. We can always integrate this as it is, and change this before 4.1 is released if we think it makes sense.

NoelDeMartin commented on src/core/classes/page-loads-manager.ts:
```diff
@@ -0,0 +1,146 @@
+            this.onRefreshPage.next();
+        }
+
+        this.removeLoadWatcher(loadWatcher);
```
I think this could be called just after the first if instead, and that way we only need to call it once and we can clean up the code a bit (with early returns and such).

NoelDeMartin commented on src/core/classes/site.ts:
```diff
@@ -2193,6 +2391,68 @@ export class CoreSite {
+                    await CoreUtils.nextTick();
+
+                    if (isCompleted) {
+                        // Current request only returns cached data. Let chained requests update in background.
```
I'm not sure if I already mentioned this, but is the `complete` callback called before `next`?  I'd say it doesn't, but if this is working I guess it does?

dpalou replied:
In general, `complete` shouldn't be called before `next` because once an observable completes it can no longer emit values. If you call `next` on an observable that has already completed nothing happens.

The implementation of `chainRequests` relies on the implementation of `WSObservable` (that's the reason why I put it in the same file). Basically `WSObservable` will:
1- If it doesn't need to update in background it will call `next` and immediately call `complete`.
2- If it needs to update in background it will call `next` with first value, and later it will call `next` with the second value and then call `complete` immediately.

This means that, in the `chainRequests` implementation, the `next` callback will always be executed before the `complete` callback (that's true for all observables), and so `subscriber.next` will always be called before `subscriber.complete`. There is a `nextTick()` wait in the `next` callback to see if `complete` was called immediately (case 1 above), that's why I added this `nextTick` in the `complete` callback too to make sure `subscriber.next` is executed first. And it works.

This might be implemented with a queue of values to emit to avoid the `nextTick` in the `complete` callback, but we'd still need a `nextTick` or `setTimeout` in the `next` callback to emit the value if `complete` wasn't called immediately so I thought this solution was simpler.

NoelDeMartin commented on src/core/classes/site.ts:
```diff
@@ -2193,6 +2391,68 @@ export class CoreSite {
+        let isCompleted = false;
+
+        return source.subscribe({
+            next: async (value) => {
```
This entire method could be refactored using early returns to reduce indentation and improve readability.

NoelDeMartin commented on src/core/features/filter/services/filter.ts:
```diff
@@ -284,13 +284,15 @@ export class CoreFilterProvider {
                 param: 'contexts',
                 maxLength: 300,
             },
+            // Use update in background, but always use the first value. If data is updated it will be stored in DB.
+            updateInBackground: true,
```
```suggestion
            ...CoreSites.getReadingStrategyPreSets(CoreReadingStrategy.STALE_WHILE_REVALIDATE),
```

What do you think of doing this instead? I'm not sure why the settings related with caching are not set anyways.

NoelDeMartin commented on src/core/services/sites.ts:
```diff
@@ -2017,6 +2023,8 @@ export const enum CoreSitesReadingStrategy {
     PREFER_CACHE,
     ONLY_NETWORK,
     PREFER_NETWORK,
+    STALE_WHILE_REVALIDATE,
+    NO_EMERGENCY_CACHE,
```
```suggestion
    STALE_WHILE_REVALIDATE,
```

NoelDeMartin commented on src/core/singletons/subscriptions.ts:
```diff
@@ -31,37 +32,46 @@ export class CoreSubscriptions {
+        const runCallback = (callback) => {
+            if (!callbackCalled) {
+                callbackCalled = true;
+                callback();
```
Why is this necessary? If an observable is resolved and completed, wouldn't it be correct to call both onSuccess and onComplete?

Looking at this, and my previous comment about what happens first complete/next, I wonder if `firstValueFrom` is correct as well. If `complete` is really called before next, it could potentially throw an error. If not, the code with my previous comment could be wrong.

dpalou replied:
IMO in `once` we only want to trigger one callback, it doesn't matter the type (next/error/complete). As I said above, once `complete` is called then the next/error ones will never be called (the same happens with `error`, once error is called then next/complete will never be called).

dpalou commented on src/addons/block/myoverview/components/myoverview/myoverview.ts:
```diff
@@ -369,13 +426,17 @@ export class AddonBlockMyOverviewComponent extends CoreBlockBaseComponent implem
     protected async refreshCourseList(data: CoreCoursesMyCoursesUpdatedEventData): Promise<void> {
         if (data.action == CoreCoursesProvider.ACTION_ENROL) {
             // Always update if user enrolled in a course.
+            this.loaded = false;
```
As we talked, I added a `showLoading` parameter in the `refreshContent` function.
